### PR TITLE
[animations] Adds CompositeAnimatedWidget to ensure shape of entering and exiting subtree are the same.

### PR DESCRIPTION
**Pull request by grinder15**
## Description
Transition widgets like FadeScaleTransition, FadeThroughTransition, and SharedAxisTransition is changing their subtree levels of widgets when transitioning causing the child state object to get recreated.

CompositeAnimatedWidget solves the issue, instead of rebuilding animation widgets in AnimatedBuilder, it builds its transitions, stacked and only changing animation values to achieve different transitions on forward and reverse.

## Issue
https://github.com/flutter/flutter/issues/53550
https://github.com/flutter/flutter/issues/55826
https://github.com/flutter/flutter/issues/55874

### CODE REVIEW

**Comment by grinder15** [comment]
@googlebot I signed it!

File: packages/animations/lib/src/fade_scale_transition.dart [hunk 1]
```diff
@@ -2,6 +2,7 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import 'package:animations/src/utils/composite_animation_widget.dart';
```
Comment (goderbauer) [inline packages/animations/lib/src/fade_scale_transition.dart:5]: this should be a relative import.

**Comment by goderbauer** [comment]
(Haven't looked at this PR in details yet) Can you please fix the failing Cirrus check: https://github.com/flutter/packages/pull/141/checks?check_run_id=602145199

**Comment by grinder15** [comment]
I just did a bit of test and I found out that the issue that I created before can be easily fixed using GlobalKeys.

https://gist.github.com/grinder15/fb908c42b5f968b89476b6216d17953d

I also checked the state of widget when transitioning in a route is also fine.
I hope this can also be include in the docs of using keys when using transition widgets.
I'm gonna close this PR and the linked issue,

File: packages/animations/lib/src/utils/composite_animation_widget.dart [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+  /// child appears.
+  ///
+  /// It's animation widgets will only play when the [animation]'s effective
+  /// [AnimationStatus] goes forward.
+  ///
+  /// The animation widgets defined here must have a visible state at
+  /// [AnimationStatus.completed] (1.0).
+  ///
+  /// Must not change subtree levels.
+  final _EnterTransitionBuilder enterTransitionBuilder;
```
Comment (goderbauer) [inline packages/animations/lib/src/utils/composite_animation_widget.dart:75]: _EnterTransitionBuilder should be public (without the underscore) if it is used in the interface of a public class.

File: packages/animations/lib/src/utils/composite_animation_widget.dart [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+  /// child disappears.
+  ///
+  /// It's animation widgets will only play when the [animation]'s effective
+  /// [AnimationStatus] goes reverse.
+  ///
+  /// The animation widgets defined here must have a visible state at
+  /// [AnimationStatus.dismissed] (0.0).
+  ///
+  /// Must not change subtree levels.
+  final _ExitTransitionBuilder exitTransitionBuilder;
```
Comment (goderbauer) [inline packages/animations/lib/src/utils/composite_animation_widget.dart:87]: same here.

File: packages/animations/lib/src/utils/composite_animation_widget.dart [hunk 4]
```diff
@@ -0,0 +1,253 @@
+// Copyright 2019 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:animations/animations.dart';
+import 'package:flutter/material.dart';
+
+/// Signature for a function that creates a widget that builds a
```
Comment (goderbauer) [inline packages/animations/lib/src/utils/composite_animation_widget.dart:8]: a -> an

File: packages/animations/lib/src/utils/composite_animation_widget.dart [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+/// enter transition.
+///
+/// Used by [CompositeAnimationWidget]
+typedef _EnterTransitionBuilder = Widget Function(
+    BuildContext, Animation<double>, Widget);
+
+/// Signature for a function that creates a widget that builds a
```
Comment (goderbauer) [inline packages/animations/lib/src/utils/composite_animation_widget.dart:15]: a -> an

File: packages/animations/lib/src/utils/composite_animation_widget.dart [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+/// exit transition.
+///
+/// Used by [CompositeAnimationWidget]
+typedef _ExitTransitionBuilder = Widget Function(
```
Comment (goderbauer) [inline packages/animations/lib/src/utils/composite_animation_widget.dart:19]: These typedefs are both the same. Just define it once with a more generic name, e.g. TransitionBuilder.

File: packages/animations/lib/src/utils/composite_animation_widget.dart [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+    BuildContext, Animation<double>, Widget);
+
+/// A widget that animate it's child with different transitions when the
```
Comment (goderbauer) [inline packages/animations/lib/src/utils/composite_animation_widget.dart:22]: it's -> its

File: packages/animations/lib/src/utils/composite_animation_widget.dart [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+/// [CompositeAnimationWidget.animation] value goes forward
```
Comment (goderbauer) [inline packages/animations/lib/src/utils/composite_animation_widget.dart:23]: You shouldn't need the `CompositeAnimationWidget.` here, dartdoc should figure this out on its own.

```suggestion
/// [animation] value goes forward
```

(here and everywhere else in this block of text)

File: packages/animations/lib/src/utils/composite_animation_widget.dart [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+/// (built by [CompositeAnimationWidget.enterTransitionBuilder])
+/// and reverse
+/// (built by [CompositeAnimationWidget.exitTransitionBuilder]).
+///
+/// It is used in [FadeThroughTransition], [SharedAxisTransition],
```
Comment (goderbauer) [inline packages/animations/lib/src/utils/composite_animation_widget.dart:28]: You can remove this sentence. It's not super-relevant.

File: packages/animations/lib/src/utils/composite_animation_widget.dart [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+/// and [FadeScaleTransition] widgets.
+///
+/// This widget also preserve it's child state by building all of it's enter
```
Comment (goderbauer) [inline packages/animations/lib/src/utils/composite_animation_widget.dart:31]: it's -> its

Comment (goderbauer) [inline packages/animations/lib/src/utils/composite_animation_widget.dart:31]: nit: this sentence is too long. Split it up to make it more readable?

File: packages/animations/lib/src/utils/composite_animation_widget.dart [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+/// causes the state object in that particular subtree level to get recreated.
+///
+/// The state preservation of the child will take effect if and only if you're
+/// not changing subtree levels in defined transition builders.
+class CompositeAnimationWidget extends StatefulWidget {
+  /// Creates a widget that animate it's child with different transitions when
+  /// the [animation] value goes forward
+  /// (built by [enterTransitionBuilder])
+  /// and reverse
+  /// (built by [exitTransitionBuilder]).
```
Comment (goderbauer) [inline packages/animations/lib/src/utils/composite_animation_widget.dart:44]:
```suggestion
Creates a widget that animates child according to [enterTransitionBuilder] when [animation] runs forward and [exitTransitionBuilder] when it runs in reverse.
```

File: packages/animations/lib/src/utils/composite_animation_widget.dart [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+  const CompositeAnimationWidget({
+    @required this.animation,
+    @required this.child,
```
Comment (goderbauer) [inline packages/animations/lib/src/utils/composite_animation_widget.dart:47]: per convention, child should always be the last parameter.

File: packages/animations/lib/src/utils/composite_animation_widget.dart [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+/// The state preservation of the child will take effect if and only if you're
+/// not changing subtree levels in defined transition builders.
+class CompositeAnimationWidget extends StatefulWidget {
+  /// Creates a widget that animate it's child with different transitions when
+  /// the [animation] value goes forward
+  /// (built by [enterTransitionBuilder])
+  /// and reverse
+  /// (built by [exitTransitionBuilder]).
+  const CompositeAnimationWidget({
+    @required this.animation,
```
Comment (goderbauer) [inline packages/animations/lib/src/utils/composite_animation_widget.dart:46]: As first parameter, please add a key parameter that you pass to super.

File: packages/animations/lib/src/utils/composite_animation_widget.dart [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+    @required this.child,
+    @required this.enterTransitionBuilder,
+    @required this.exitTransitionBuilder,
+    this.flip = false,
+    this.syncVisibleStates = false,
+  })  : assert(child != null),
+        assert(animation != null),
```
Comment (goderbauer) [inline packages/animations/lib/src/utils/composite_animation_widget.dart:53]: Presumably, flip and syncVisibleStates also can't be null? Please add asserts.

File: packages/animations/lib/src/utils/composite_animation_widget.dart [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+        assert(enterTransitionBuilder != null),
```
Comment (goderbauer) [inline packages/animations/lib/src/utils/composite_animation_widget.dart:54]: The doc comment of the constructor should also document which arguments cannot be null.

File: packages/animations/lib/src/utils/composite_animation_widget.dart [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+
+  /// The animation that drives the [child]'s entrance and exit.
+  final Animation<double> animation;
+
+  /// The widget below this widget in the tree.
+  ///
+  /// This widget will transition in and out as driven by [animation].
+  final Widget child;
+
+  /// A function that wraps [child] with an animation set define how the
```
Comment (goderbauer) [inline packages/animations/lib/src/utils/composite_animation_widget.dart:65]:
```suggestion
  /// A function that wraps [child] with an animation defining how the
```

File: packages/animations/lib/src/utils/composite_animation_widget.dart [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+  /// child appears.
+  ///
+  /// It's animation widgets will only play when the [animation]'s effective
```
Comment (goderbauer) [inline packages/animations/lib/src/utils/composite_animation_widget.dart:68]: it's -> its

File: packages/animations/lib/src/utils/composite_animation_widget.dart [hunk 18]
```diff
@@ -0,0 +1,253 @@
...
+  /// [AnimationStatus] goes forward.
```
Comment (goderbauer) [inline packages/animations/lib/src/utils/composite_animation_widget.dart:69]:
```suggestion
  /// [AnimationStatus] goes forward from 0.0 to 1.0.
```

File: packages/animations/lib/src/utils/composite_animation_widget.dart [hunk 19]
```diff
@@ -0,0 +1,253 @@
...
+  ///
+  /// Flipping animation values will also flips visible states in defined
+  /// transition builders.
+  ///
+  /// So when flipped, The visible state of widgets in [enterTransitionBuilder]
+  /// and [exitTransitionBuilder] will be at [AnimationStatus.dismissed] (0.0)
+  /// and [AnimationStatus.completed] (1.0) respectively.
+  final bool flip;
+
+  /// Sync visible state of [exitTransitionBuilder] to [enterTransitionBuilder].
```
Comment (goderbauer) [inline packages/animations/lib/src/utils/composite_animation_widget.dart:99]: not sure what you mean by this based on the doc.

File: packages/animations/lib/src/utils/composite_animation_widget.dart [hunk 20]
```diff
@@ -0,0 +1,253 @@
...
+  /// It's animation widgets will only play when the [animation]'s effective
+  /// [AnimationStatus] goes reverse.
+  ///
+  /// The animation widgets defined here must have a visible state at
+  /// [AnimationStatus.dismissed] (0.0).
+  ///
+  /// Must not change subtree levels.
+  final _ExitTransitionBuilder exitTransitionBuilder;
+
+  /// Indicates whether the animation values will be flipped.
```
Comment (goderbauer) [inline packages/animations/lib/src/utils/composite_animation_widget.dart:89]: Based on the doc it's not clear to me what this is supposed to do.

File: packages/animations/lib/src/utils/composite_animation_widget.dart [hunk 21]
```diff
@@ -0,0 +1,253 @@
...
+    _setStartingAnimationValues(widget.flip);
+  }
+
+  void _animationListener(AnimationStatus animationStatus) {
+    _effectiveAnimationStatus = _calculateEffectiveAnimationStatus(
+      lastEffective: _effectiveAnimationStatus,
+      current: animationStatus,
+    );
+    // change the animation values based on effectiveAnimationStatus.
+    _setStartingAnimationValues(widget.flip);
```
Comment (goderbauer) [inline packages/animations/lib/src/utils/composite_animation_widget.dart:134]: only call this if _effectiveAnimationStatus has changed from the previous value?

File: packages/animations/lib/src/utils/composite_animation_widget.dart [hunk 22]
```diff
@@ -0,0 +1,253 @@
...
+  }
+
+  // This is called at [initState], [didUpdateWidget], and [_animationListener]
+  void _setStartingAnimationValues(bool flip) {
+    final Animation<double> baseAnimation =
```
Comment (goderbauer) [inline packages/animations/lib/src/utils/composite_animation_widget.dart:139]: this method is hard to follow because _enterAnimation and _exitAnimation change their value fairly often. Can you refactor this to make the flow clearer?

E.g. maybe start with the switch and then in each case configure _enterAnimation/_exitAnimation based on the value of flip and syncVisibleStates?

File: packages/animations/lib/src/utils/composite_animation_widget.dart [hunk 23]
```diff
@@ -0,0 +1,253 @@
...
+
+  @override
+  void didUpdateWidget(CompositeAnimationWidget oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (oldWidget.animation != widget.animation) {
+      oldWidget.animation.removeStatusListener(_animationListener);
+      widget.animation.addStatusListener(_animationListener);
+      _animationListener(widget.animation.status);
+    }
+    if (widget.flip != oldWidget.flip) {
```
Comment (goderbauer) [inline packages/animations/lib/src/utils/composite_animation_widget.dart:181]: what if syncVisibleStates changes?

File: packages/animations/lib/src/utils/composite_animation_widget.dart [hunk 24]
```diff
@@ -0,0 +1,253 @@
...
+      _setStartingAnimationValues(widget.flip);
+    }
+  }
+
```
Comment (goderbauer) [inline packages/animations/lib/src/utils/composite_animation_widget.dart:185]: This class is missing a dispose method where you remove the listener.

**Comment by goderbauer** [comment]
Some high-level feedback: This is pretty cool!

Maybe we should simplify the CompositeAnimatedWidget a little bit by removing the `flip` and `syncVisibleStates` flag. Instead, we define that the animations given to the enterBuilder and exitBuilder will always run from 0.0 to 1.0. The enterBuilder is supposed to make the widget appear as its animation runs from 0.0 to 1.0 and the exit builder is supposed to make the animation disappear as it runs from 0.0 to 1.0. We then adjust the individual enter and exit Builder passed into the CompositeAnimatedWidget by e.g. FadeScaleTransition, FadeThroughTransition, and SharedAxisTransition to match this model.

Last, but not least, can you also add tests to this PR please? Otherwise we will not be able to merge it as per our policy.

File: packages/animations/lib/src/utils/composite_animation_widget.dart [hunk 25]
```diff
@@ -0,0 +1,253 @@
+// Copyright 2019 The Flutter Authors. All rights reserved.
```
Comment (goderbauer) [inline packages/animations/lib/src/utils/composite_animation_widget.dart:1]: this widget should actually directly live in the src directory (not under utils) and we should export it from `animations.dart`. I think this class can be handy for other use cases as well.

File: packages/animations/lib/src/utils/composite_animation_widget.dart (hunk 2 above)
Comment (grinder15) [reply packages/animations/lib/src/utils/composite_animation_widget.dart:75]: got it

File: packages/animations/lib/src/utils/composite_animation_widget.dart (hunk 3 above)
Comment (grinder15) [reply packages/animations/lib/src/utils/composite_animation_widget.dart:87]: got it

File: packages/animations/lib/src/utils/composite_animation_widget.dart (hunk 4 above)
Comment (grinder15) [reply packages/animations/lib/src/utils/composite_animation_widget.dart:8]: got it

File: packages/animations/lib/src/utils/composite_animation_widget.dart (hunk 5 above)
Comment (grinder15) [reply packages/animations/lib/src/utils/composite_animation_widget.dart:15]: got it

File: packages/animations/lib/src/utils/composite_animation_widget.dart (hunk 6 above)
Comment (grinder15) [reply packages/animations/lib/src/utils/composite_animation_widget.dart:19]: got it

File: packages/animations/lib/src/utils/composite_animation_widget.dart (hunk 20 above)
Comment (grinder15) [reply packages/animations/lib/src/utils/composite_animation_widget.dart:89]: I just point it out to define builders without conditional building

File: packages/animations/lib/src/utils/composite_animation_widget.dart (hunk 21 above)
Comment (grinder15) [reply packages/animations/lib/src/utils/composite_animation_widget.dart:134]: This is called in `initState`, `didUpdateWidget`, and `_animationListener`
It changes the internal `Animation` object's values based on `flip` flag, `syncVisibleStates` and `_effectiveAnimationStatus`.

File: packages/animations/lib/src/utils/composite_animation_widget.dart (hunk 22 above)
Comment (grinder15) [reply packages/animations/lib/src/utils/composite_animation_widget.dart:139]: Ok, And I think ill make an additional `_setStartingAnimationValues` function parameters for `effectiveAnimationStatus` and `syncVisibleState` or access them directly to widget property?

File: packages/animations/lib/src/utils/composite_animation_widget.dart (hunk 24 above)
Comment (grinder15) [reply packages/animations/lib/src/utils/composite_animation_widget.dart:185]: got it

File: packages/animations/lib/src/utils/composite_animation_widget.dart (hunk 25 above)
Comment (grinder15) [reply packages/animations/lib/src/utils/composite_animation_widget.dart:1]: got it

File: packages/animations/lib/src/utils/composite_animation_widget.dart (hunk 23 above)
Comment (grinder15) [reply packages/animations/lib/src/utils/composite_animation_widget.dart:181]: The default flow of values in those animations objects if `syncVisibleStates` is false
_enterAnimation: 0.0  -> 1.0
_exitAnimation: 1.0 -> 0.0

in `syncVisibleState` is true

_enterAnimation: 0.0  -> 1.0
_exitAnimation: 0.0 -> 1.0 (syncs to enterAnimation, same visibleState)

just to clarify the _exitAnimation is always flip against _enterAnimation and all "End" animation values are always the visible states, "flip" flag will flip all of the values mentioned above.

the only usage of `syncVisibleStates` set to true is in `FadeScaleTransition` where in the original implementation the exit animation is not flipped.

**Comment by grinder15** [comment]
@godofredoc Thanks for the review, I'll apply those changes this coming weekend. No idea though to how to do a test, I think I'll start with `CompositeAnimationWidget` ensuring animation values 0.0 to 1.0 in both builders, any thoughts? hints?

First time to contribute, so please be nice to me, :sweat_smile: Thanks!

**Comment by shihaohong** [comment]
@grinder15 I think you meant to tag @goderbauer :)

File: packages/animations/lib/src/composite_animation_widget.dart [hunk 26]
```diff
@@ -0,0 +1,270 @@
...
+///
+/// To ensure the state of the child widget will survive, don't define
+/// conditional logic to switch subtrees in both transition builders required
+/// by this widget.
+class CompositeAnimationWidget extends StatefulWidget {
+  /// Creates a widget that animate it's child with different transitions based
+  /// on [AnimationStatus] of [animation].
+  ///
+  /// If the ([animation]) value goes forward, only [forwardTransitionBuilder]
+  /// animates from 0.0 to 1.0 and If goes reverse, only [reverseTransitionBuilder]
```
Comment (goderbauer) [inline packages/animations/lib/src/composite_animation_widget.dart:58]: I think it will be easier to reason about the animations if they go from 0.0 to 1.0 for both builders. Let's say you want to change the background from green to white on exit. If animations go from 0.0 to 1.0 all you have to do is define a `Tween(begin: Colors.green, end: Colors.white)`. If they go from 1.0 to 0.0 you have to switch begin/end around, which is awkward to think about. (This will require some changes to the existing animations, e.g. begin/end for some tweens need to be switched around and some curves may need to be flipped).

File: packages/animations/lib/src/composite_animation_widget.dart [hunk 27]
```diff
@@ -0,0 +1,270 @@
...
+  /// [reverseTransitionBuilder]) to have a
+  /// "visible-point" of animation at 0.0 ([AnimationStatus.dismissed]).
+  ///
+  /// This is usually set to true if you want to animate a widget to
+  /// disappear but with different exit and reenter animation without enter
+  /// animation (visible at start), which means you defined exit animation at
+  /// [forwardTransitionBuilder] and reenter animation
+  /// at [reverseTransitionBuilder].
+  ///
+  /// Some transition implementations like [FadeThroughTransition] and
```
Comment (goderbauer) [inline packages/animations/lib/src/composite_animation_widget.dart:109]: It's not entirely clear to me why we need this setting for the FadeThroughTransition and SharedAxisTransition. The FadeThroughTransition and SharedAxisTransition should just define their forwardTransitionBuilder/reverseTransitionBuilder properly.

**Comment by goderbauer** [comment]
>  I would also add a test for each of the FadeScaleTransition, FadeThroughTransition, and SharedAxisTransitions that now use the CompositeAnimatedWidget to make sure that state is not lost after the transition animations complete in forwards, or in reverse.

I also like @shihaohong's suggestion of adding a test for this.

**Comment by shihaohong** [comment]
I tried to write a test for the SharedAxisTransition case: https://github.com/grinder15/packages/pull/1

It seems to be a problem if we ever use the transition in tandem with the `PageTransitionSwitcher`. I wonder if there's a more general solution that can be applied to `PageTransitionSwitcher` and not have to use the `CompositeAnimatedWidget` for every transition that is ever used with `PageTransitionSwitcher`. For example, using `PageTransitionsBuilder` with `SharedAxisTransitionsPageBuilder` does not cause the problem because the widget tree shape does not alter.

File: packages/animations/lib/src/composite_animation_widget.dart (hunk 27 above)
Comment (grinder15) [reply packages/animations/lib/src/composite_animation_widget.dart:109]: To better understanding, lets take a look at this table
compo
The use of `visibleAtStart` explicitly telling the widget that we want to define an "exit"/"disappear" animation, thus the secondary animation part of some page transition widgets. This is also the reason why I renamed those builders to avoid confusion like defining an "exit" animation in `enterBuilder` and "reenter" animation in `exitBuilder`.

On the technical side, we want the widget to know where those "values" stays while the another builder is animating. If we don't set the flag correctly, it will end up the builder to animate but we cant see the actual animation since the all the animation widget defined are built by composition being the `forwardBuilder` is the parent and the child is the `reverseBuilder`. For example, if the `reverseBuilder`'s `FadeTransition` opacity stays at '0.0' and the `forwardBuilder`'s `ScaleTransition` is animating, we can't see the animation. 

```
return widget.forwardTransitionBuilder(
          widget.reverseTransitionBuilder(
            child,
            _reverseAnimation,
          ),
          _forwardAnimation,
        );
```

`visibleAtStart` sets the x values to `0.0` if true, else '1.0'.

With your suggestion about `reverseBuilder` must run forward, I can simply add a flip function to `_reverseAnimation` .

**Comment by grinder15** [comment]
@shihaohong Thanks for the sample test, I will add these with my later commits.

About the animation problem, I also tried to use page transition widgets to hook it up directly to `AnimationController`, the same problem occurs with stateful widgets.
I think it is good to think those page transition widgets has the same use cases as the plain old transition widgets in flutter (`ScaleTransition`, `FadeTransition`, etc.) like respond to drag events, adding animation in `AnimatedList`, or animate when it is mounted in ScrollView.

English is not my primary language but I hope you get my point :)

**Comment by shihaohong** [comment]
@grinder15 I think I sort of get what you mean, but maybe an easy way to demonstrate is with a short code sample? 

From what I'm reading, it sounds like if you were to directly pass in your own `AnimationController` into, for example, `FadeScaleTransition.animation`, and drive the animation forwards/backwards programmatically, you would still get the same issue?

**Comment by grinder15** [comment]
@shihaohong 
https://gist.github.com/grinder15/e3cc69ad82a545407f0eaa2c28355e65
And it goes like this..
anim
As you transition in and out, it resets the counter widget which is suppose to persist.

**Comment by shihaohong** [comment]
@grinder15 That makes sense, thanks for sharing the demo!

**Comment by goderbauer** [comment]
@grinder15 Inspired by your work I took a stab at this in https://github.com/flutter/packages/pull/160. My main change there is that I eliminated the need for the `visibleAtStart` flag, which I think makes the widget easier to use and understand. Instead, I am adapting the animation that gets passed into the widget to match the expected model. I also changed some of the names. Let me know what you think about that approach and whether I am missing a use case that doesn't work in my approach.

**Comment by shihaohong** [comment]
Since https://github.com/flutter/packages/pull/160 has been merged, I will now close this issue. @grinder15 Thank you for the initial solution and investigation! I think this was a really cool idea and helped shape a much cleaner solution to the problem

**Comment by xhowe** [comment]
> @shihaohong
> https://gist.github.com/grinder15/e3cc69ad82a545407f0eaa2c28355e65
> And it goes like this..
> anim
> As you transition in and out, it resets the counter widget which is suppose to persist.

Meet the same problem，how are things going？

**Comment by grinder15** [comment]
@xhowe This was already solved in the latest version. Try to update your pubspec.yaml
